Path: Fix extensions updates and Default Length application [Bug]

Russ4262 opened the pull request:
These changes restore the extension size updating mechanism due to upstream code changes.
These changes fix the annoying requirement to apply Default Length value changes via a click to the task panel or viewport, by adding a manual signal emission.

Forum discussion at [Bug: Extensions - default length edit not updating extension sizes](https://forum.freecadweb.org/viewtopic.php?f=15&t=67699).

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [x]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [x]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_6703/pipeline.svg"/></a> for feature branch [PR_6703](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703). Pipeline [507572750](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/507572750) was triggered at [1d708bc](https://github.com/FreeCAD/FreeCAD/pull/6703/commits/1d708bc8eed25e343e778a508c715dc2a9378a2b). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

mlampert commented on src/Mod/Path/PathScripts/PathFeatureExtensionsGui.py:
```diff
@@ -299,13 +301,11 @@ def _initializeExtensions(self, obj):
+        """_updateExtensionSizes(index=None)...
+        Slot method to update extensions due to change in Default Length."""
+        if self.form.defaultLength.text() != self.lastDefaultLength:
+            self.lastDefaultLength = self.form.defaultLength.text()
```
Are you sure you don't have to call `self.defaultLength.updateSpinBox()` anymore?

Russ4262 replied:
Updates are being done by done by the event handler in the `QuantitySpinBox` class of the PathGui module.  Anytime the event is `FocusIn` it is calling it.  However, I was unsure on this one.  My tests show that values are being passed as needed for the extensions, and reloaded correctly when editing.  When I left it in, I was seeing double updates of the SpinBox, so I took it out.   I will place it back, just to be sure I am not messing something up in the background that I am not seeing.

Thanks for the review!

Russ4262 replied:
Okay.  Just added back the `updateSpinBox()` call.

mlampert commented on src/Mod/Path/PathScripts/PathGui.py:
```diff
@@ -129,6 +129,7 @@ def __init__(self, widget, obj, prop, onBeforeChange=None):
     def eventFilter(self, obj, event):
         if event.type() == QtCore.QEvent.Type.FocusIn:
             self.updateSpinBox()
+            self.widget.editingFinished.emit()
```
`FocusIn` is the event when the widget _gets_ the focus which is not the time to emit `editingFinished`

Russ4262 replied:
The widget is the SpinBox, not the QLineEdit for the expression, so yes, we need it after the expression editing session is finished with the QLineEdit and focus is returned back to the SpinBox.   If I move the signal emission to `Type == FocusOut`, then we are back where we started with the user having to manually click on the task panel or view port for the code to apply the change to Default Length.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_6703/pipeline.svg"/></a> for feature branch [PR_6703](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703). Pipeline [507667679](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/507667679) was triggered at [7d664b6](https://github.com/FreeCAD/FreeCAD/pull/6703/commits/7d664b65d9303ce2dd9396e67f7d431e097a50d3). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

Russ4262 commented:
Tests are showing this is behaving as intended, both with updating an expression, and with just the spinbox when the expression has been cleared, and then back again to an expression.

mlampert commented:
It might behave as you intended for this scenario, but it's not a "fix". Issuing an `editingFinished` signal when something _gets_ the focus is not correct.

Russ4262 commented:
@mlampert Will you please help me make a custom signal that I can connect to a slot?

Russ4262 commented:
@mlampert 
Thanks for your feedback and assistance.

mlampert commented:
This might be an issue in the QuantitySpinBox itself, which is above my pay grade. Maybe post in the developer forum, someone there might know what's going on and what caused this regression.

Russ4262 commented:
[Qt via PySide: Expression editor back to SpinBox event or signal?](https://forum.freecadweb.org/viewtopic.php?f=10&t=67732)

Thanks for your kindness and patience.  I very much appreciate your pay grade. It is quite helpful.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_6703/pipeline.svg"/></a> for feature branch [PR_6703](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703). Pipeline [508084052](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/508084052) was triggered at [3392995](https://github.com/FreeCAD/FreeCAD/pull/6703/commits/339299539efd57ad0ec4f811c58c032a6a6590b4). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

Russ4262 commented:
@mlampert Evening.  I think I have it!  I just rebased and updated the PR.  The `editingFinished` signal does need to be emitted (best I can tell) to achieve the immediate update, but as you say, it does not need to be emitted in the location of the previous code (in the event handler of the QuantitySpinBox class), as that would likely introduce possible errors in other usages throughout Path, if not now, eventually. 

I believe I have the solution.  What is missing is intermediate code to determine if the Default Length value change is due to an expression edit, or a simple spinbox edit (when there is no expression engine used).  This is the code I added.  It is working as expected!   Thanks for pointing out the flaw in my earlier version and holding firm.

I think you will find this version more suitable.  Have a great night!

mlampert commented on src/Mod/Path/PathScripts/PathFeatureExtensionsGui.py:
```diff
@@ -299,16 +301,33 @@ def _initializeExtensions(self, obj):
-        prevValue = self.form.defaultLength.text()
-        self.defaultLength.updateSpinBox()
-        postValue = self.form.defaultLength.text()
+    def _isDefaultLengthExpression(self):
```
nitpick: the name here is a bit confusing because it's not clear what _is_ the expression. I think it would read and could be processed easier if you would name it `_haveDefaultLengthExpression(self):`  or `_isDefaultLengthExpressionAssigned` - something along those lines.

Russ4262 replied:
Made some improvements.  I moved the `valueChanged` signal translation into the QuantitySpinBox class in the PathGui module, as that is where it belongs.  This will fix each instance where the QSB class is deployed.  It also simplified the overall fix for this issue related to extensions updating.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_6703/pipeline.svg"/></a> for feature branch [PR_6703](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703). Pipeline [509860876](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/509860876) was triggered at [aa69fc2](https://github.com/FreeCAD/FreeCAD/pull/6703/commits/aa69fc25296aeb325c6d15ca5ec6d72e7f78e981). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_6703/pipeline.svg"/></a> for feature branch [PR_6703](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703). Pipeline [509863640](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/509863640) was triggered at [f61c172](https://github.com/FreeCAD/FreeCAD/pull/6703/commits/f61c172b32c8ea9f1ed8c4e84aefd12b3bf3d10e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

Russ4262 commented on src/Mod/Path/PathScripts/PathFeatureExtensionsGui.py:
```diff
@@ -299,16 +301,33 @@ def _initializeExtensions(self, obj):
+            self._isDefaultLengthExpression()
+            and self.form.defaultLength.text() != self.lastDefaultLength
+        ):
+            self.form.defaultLength.editingFinished.emit()
```
Per Qt documentation, the `editingFinished` signal is handled differently than the `valueChanged`.  The `editingFinished` is that same signal that is emitted in a regular QLineEdit when the user has finished editing text and wants the focus to return to the parent widget or task panel.  That is what we need in this case for the extension update code to properly update.  Without the `editingFinished` the code does not update the extensions because focus is still on the QSB - which is why we have had to click outside the QSB after using the Forumula Editor window successfully.

As posted in above code review response, I have incorporated this `valueChanged` to `editingFinished` signal translation into the QuantitySpinBox class where it belongs.  I tested the `StartDepth` inputs (another QSB usage) a number of times and they are behaving correctly and accordingly with these changes.

Thanks for your review and comments and suggestions!

Russ4262 commented:
Just rebased on master.  No changes.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_6703/pipeline.svg"/></a> for feature branch [PR_6703](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6703). Pipeline [533569407](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/533569407) was triggered at [03c5661](https://github.com/FreeCAD/FreeCAD/pull/6703/commits/03c56616907bd6fc75f614baead7be3dc0427e6f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

sliptonic commented:
We need to decide whether to merge, delay to 0.21, or add to the 0.20 milestone.  Please comment

Russ4262 commented:
This one is ready for merge from my end, but I do defer to MLampert and other devs.
